dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -351,140 +530,219 @@ def _get_outputs(self, pipe_run_ref: DALIPipelineRunRef):
+        if req_batch_id in self._cache_outputs:
+            req_outputs = self._cache_outputs[req_batch_id]
+            del self._cache_outputs[req_batch_id]
 
         else:
             curr_batch_id = None
             # If not the data we are looking for, store it and keep processing until we find it
             while req_batch_id != curr_batch_id:
-                torch.cuda.nvtx.range_push("dali_output_q.get")
-                curr_batch_id, curr_processed_outputs = self.dali_output_q.get()
-                torch.cuda.nvtx.range_pop()
+                _nvtx.range_push("dali_output_q.get")
+                curr_batch_id, curr_processed_outputs, err = self._dali_output_q.get()
+                _nvtx.range_pop()
+
+                if err is not None:
+                    raise err
+
                 if curr_batch_id == req_batch_id:
                     req_outputs = curr_processed_outputs
                 else:
-                    self.cache_outputs[curr_batch_id] = curr_processed_outputs
-        # Unpack single element tuples
-        if isinstance(req_outputs, tuple) and len(req_outputs) == 1:
-            req_outputs = req_outputs[0]
+                    self._cache_outputs[curr_batch_id] = curr_processed_outputs
         return req_outputs
 
+    def _produce_data_impl(self, obj, cache):
+        """
+        Recursive implementation of produce_data
+        """
+
+        # Check if the object has already been visited (to prevent infinite recursion)
+        obj_id = id(obj)
+        if obj_id in cache:
+            return cache[obj_id]
+
+        def convert():
+            # DALIOutputBatchRef instance, replace it with data
+            if isinstance(obj, DALIOutputBatchRef):
+                pipe_run_ref_id = id(obj.pipe_run_ref)
+                if pipe_run_ref_id not in cache:
+                    cache[pipe_run_ref_id] = self._get_outputs(obj.pipe_run_ref)
+                return cache[pipe_run_ref_id][obj.output_idx]
+
+            # custom class, recursively call produce data on its members
+            elif hasattr(obj, "__dict__"):
+                cache[obj_id] = obj  # prevent infinite recursion
+                for attr_name, attr_value in obj.__dict__.items():
+                    setattr(obj, attr_name, self._produce_data_impl(attr_value, cache))
+                return obj
+
+            # list, recursively apply the function to each element
+            elif isinstance(obj, list):
+                ret_list = []
+                cache[obj_id] = ret_list  # prevent infinite recursion
+                for item in obj:
+                    ret_list.append(self._produce_data_impl(item, cache))
+                return ret_list
+
+            # tuple, recursively apply the function to each element (preserving tuple type)
+            elif isinstance(obj, tuple):
+                # tuple is immutable (can't contain circular references)
+                return tuple(self._produce_data_impl(item, cache) for item in obj)
```

Comment:
Btw, I realized that if we are dealing with potentially custom, user provided types, we may break stuff. This potentially goes back to the tuple or list as a base class if user inherited from them.
```
class A(list):
    pass

isinstance(A(), list) # True
```
So we probably need to recreate the object as `type(obj)(...)`, but the user might have done something weird with the constructor.

Reply:
fixed